From: Ben Hutchings <ben@decadent.org.uk>
Subject: rtnetlink: avoid ABI change in 3.2.34
Date: Sun, 18 Nov 2012 03:25:48 +0000
Forwarded: not-needed

Commit d318a127e273716c9531fe70d497ca24db4c0bf1 ('rtnetlink: Fix
problem with buffer allocation', commit
115c9b81928360d769a76c632bae62d15206a94a upstream) added a parameter
to the optional 'calcit' callback for rtnetlink operations.  There is
only one in-tree implementation and it's in the same file as the
caller, so call it directly with the extra argument when necessary.

Also, the ifla_policy array grew, but this is backward-compatible so
hide the change from genksyms.

---
Index: linux-3.2.46/include/net/rtnetlink.h
===================================================================
--- linux-3.2.46.orig/include/net/rtnetlink.h	2013-07-26 18:54:02.000000000 +0000
+++ linux-3.2.46/include/net/rtnetlink.h	2013-07-26 19:25:10.000000000 +0000
@@ -6,7 +6,7 @@
 
 typedef int (*rtnl_doit_func)(struct sk_buff *, struct nlmsghdr *, void *);
 typedef int (*rtnl_dumpit_func)(struct sk_buff *, struct netlink_callback *);
-typedef u16 (*rtnl_calcit_func)(struct sk_buff *, struct nlmsghdr *);
+typedef u16 (*rtnl_calcit_func)(struct sk_buff *);
 
 extern int	__rtnl_register(int protocol, int msgtype,
 				rtnl_doit_func, rtnl_dumpit_func,
Index: linux-3.2.46/net/core/rtnetlink.c
===================================================================
--- linux-3.2.46.orig/net/core/rtnetlink.c	2013-07-26 18:54:02.000000000 +0000
+++ linux-3.2.46/net/core/rtnetlink.c	2013-07-26 19:25:10.000000000 +0000
@@ -1117,7 +1117,9 @@
 	[IFLA_VF_PORTS]		= { .type = NLA_NESTED },
 	[IFLA_PORT_SELF]	= { .type = NLA_NESTED },
 	[IFLA_AF_SPEC]		= { .type = NLA_NESTED },
+#ifndef __GENKSYMS__
 	[IFLA_EXT_MASK]		= { .type = NLA_U32 },
+#endif
 };
 EXPORT_SYMBOL(ifla_policy);
 
@@ -2021,7 +2023,9 @@
 			return -EOPNOTSUPP;
 		calcit = rtnl_get_calcit(family, type);
 		if (calcit)
-			min_dump_alloc = calcit(skb, nlh);
+			min_dump_alloc = calcit(skb);
+		else if (type == RTM_GETLINK)
+			min_dump_alloc = rtnl_calcit(skb, nlh);
 
 		__rtnl_unlock();
 		rtnl = net->rtnl;
@@ -2137,7 +2141,7 @@
 	register_netdevice_notifier(&rtnetlink_dev_notifier);
 
 	rtnl_register(PF_UNSPEC, RTM_GETLINK, rtnl_getlink,
-		      rtnl_dump_ifinfo, rtnl_calcit);
+		      rtnl_dump_ifinfo, NULL);
 	rtnl_register(PF_UNSPEC, RTM_SETLINK, rtnl_setlink, NULL, NULL);
 	rtnl_register(PF_UNSPEC, RTM_NEWLINK, rtnl_newlink, NULL, NULL);
 	rtnl_register(PF_UNSPEC, RTM_DELLINK, rtnl_dellink, NULL, NULL);
Index: linux-3.2.46/include/linux/if_link.h
===================================================================
--- linux-3.2.46.orig/include/linux/if_link.h	2013-07-26 18:54:02.000000000 +0000
+++ linux-3.2.46/include/linux/if_link.h	2013-07-26 19:25:10.000000000 +0000
@@ -137,7 +137,9 @@
 	IFLA_AF_SPEC,
 	IFLA_GROUP,		/* Group the device belongs to */
 	IFLA_NET_NS_FD,
+#ifndef __GENKSYMS__
 	IFLA_EXT_MASK,		/* Extended info mask, VFs, etc */
+#endif
 	__IFLA_MAX
 };
 
